### ISSUE TITLE
Loss settings when training with real Aloha robot dataset

### ISSUE THREAD

**Issue by a510721** [mlzxy/arp#11]
Thank you for your kind response all the time.
If I want to proceed with learning based on the real aloha robot dataset, only the values of joint, action and camera images are saved. 
In simulation, we calculate action.nll_loss, guide-pt-left.2d_ce_loss, and guide-pt-right.2d_ce_loss. If I train using only action.nll_loss from the real robot dataset, will I get the same result?
Can I exclude only guide-pt-left.2d_ce_loss and guide-pt-right.2d_ce_loss? 
I would be glad if you could also provide an explanation of loss(action.nll_loss, guide-pt-left.2d_ce_loss, and guide-pt-right.2d_ce_loss)

**Comment by mlzxy** [comment]
> In simulation, we calculate action.nll_loss, guide-pt-left.2d_ce_loss, and guide-pt-right.2d_ce_loss. If I train using only action.nll_loss from the real robot dataset, will I get the same result?  

If you don't use the guide-pt loss, the overall performance will be lower than ACT. The guide-pt are 2d waypoints of the end-effectors. By predicting them in autoregression, the model learns better action prediction. 

I uploaded an example in [2d-waypoints-real-robot.ipynb](https://github.com/mlzxy/arp/blob/main/real-robot/2d-waypoints-real-robot.ipynb) to show how to compute 2d waypoints. Suppose you have robot joint values, the examples show you how to (1) use forward kinematics to compute robot joint 3d xyz (2) use camera intrinsics+extrinsics to compute the robot joint 2d xyz, which can then be selected / subsampled as 2d waypoints.  

We also have [aloha/compute_waypoints.py](https://github.com/mlzxy/arp/blob/main/aloha/compute_waypoints.py). But this script obtains the 2d waypoints by directly communicating with the mujoco environment. 

---

> I would be glad if you could also provide an explanation of loss(action.nll_loss, guide-pt-left.2d_ce_loss, and guide-pt-right.2d_ce_loss)
> I'm training with action.null_loss, but the loss value is negative. Is it okay? what does 'minus' mean?

The action.nll_loss is the negative log likelihood loss ([L781](https://github.com/mlzxy/arp/blob/70252f02d43628d7395333db3c073b58c930b702/arp.py#L781)), that mixmizing the likelihood of continuous actions in a mixtured-gaussian distribution. So yes, it can be negative, which means very high likelihood. 

The guide ce losses are the cross entropy loss that applys on the 2d pixel coordinate actions. ([L982](https://github.com/mlzxy/arp/blob/70252f02d43628d7395333db3c073b58c930b702/arp.py#L982))  

All these losses are basically maximizing the likelihood of given action sequences. 

---

> Can I exclude only guide-pt-left.2d_ce_loss and guide-pt-right.2d_ce_loss?  
> In creating arp.TokenType.make for guide_token_right and guide_token_left, I set is_control=True to prevent them from being calculated in the loss.  

Don't set them to `is_control=True`. The `is_control=True` tokens are designed for cases like language prompts, basically the information that already present before robot execution.  Guide 2d points are still needed to be predicted.

**Comment by a510721** [comment]
The action command performs regression based on gaussian_nll_loss, while guide-pt-left and guide-pt-right approach the point location in the image as a classification problem to calculate the overall loss.

Should I calculate guide-pt-left and guide-pt-right even when only robot grippers are visible in cam_left_wrist and cam_right_wrist, but the entire robot is not seen in the image?
https://mobile-aloha.github.io/

**Comment by mlzxy** [comment]
Hi @a510721. You can truncate them to be within the image.  I draw an image to illustrate the idea, the red dots are right arm waypoint, and green dots are left arm waypoints, the data is screenshot from https://drive.google.com/drive/folders/17r6CBQKgXEt_0DfvBrSJkyZQMi66TAz9. 

![image](https://github.com/user-attachments/assets/797c0cd3-65ed-4132-89b7-315fdb253bb4)

In this case, suppose we want to grasp the bottle, both arm needs to move downwards, which means the trajectory go out of the image. You can clip them to be within the image, so they still provide a sense of direction. 

Note that, by predicting and visualizing the 2d waypoints during training, you get a good sense of whether model "understand" how to solve the tasks.

**Comment by a510721** [comment]
Is it correct that your idea is to use the robot arm trajectory only in the visible part of the image?
It seems that the trajectory of the arm joint is discrete(guide-pt-left and guide-pt-right), appearing and disappearing in the images
If I'm misunderstanding, please explain in detail.

**Comment by mlzxy** [comment]
> Is it correct that your idea is to use the robot arm trajectory only in the visible part of the image? If I'm misunderstanding, please explain in detail.

Not exact, I suggest to clip your 2d waypoints within the image, like below: 

```python
traj2d  # array of Nx2, where N is the number of waypoints, each waypoint in format x,y
# suppose your image size is h x w
traj2d[:, 0] = np.clip(traj2d[:, 0], 0, w - 1)
traj2d[:, 1] = np.clip(traj2d[:, 1], 0, h - 1)
```
